core/src/worker/slot_supplier/resource_based.rs

```diff
@@ -0,0 +1,423 @@
+    last_slot_issued_rx: watch::Receiver<Instant>,
+    last_metric_emission: AtomicCell<Instant>,
+    _slot_kind: PhantomData<SK>,
+}
+struct PidControllers {
+    mem: pid::Pid<f64>,
+    cpu: pid::Pid<f64>,
+}
+struct MetricInstruments {
+    attribs: MetricAttributes,
+    mem_usage: Arc<dyn GaugeF64>,
+    cpu_usage: Arc<dyn GaugeF64>,
+    mem_pid_output: Arc<dyn GaugeF64>,
+    cpu_pid_output: Arc<dyn GaugeF64>,
+}
+
+impl ResourceBasedSlots<RealSysInfo> {
+    /// Create an instance attempting to target the provided memory and cpu thresholds as values
+    /// between 0 and 1.
+    pub fn new(target_mem_usage: f64, target_cpu_usage: f64) -> Self {
+        Self {
+            metrics: OnceLock::new(),
+            target_mem_usage,
+            target_cpu_usage,
+            sys_info_supplier: RealSysInfo::new(),
+        }
+    }
+}
+
+impl PidControllers {
+    fn new(mem_target: f64, cpu_target: f64) -> Self {
+        let mut mem = pid::Pid::new(mem_target, 100.0);
+        mem.p(5.0, 100).i(0.0, 100).d(1.0, 100);
+        let mut cpu = pid::Pid::new(cpu_target, 100.0);
+        cpu.p(5.0, 100.).i(0.0, 100.).d(1.0, 100.);
+        Self { mem, cpu }
+    }
+}
+
+impl MetricInstruments {
+    fn new(meter: TemporalMeter) -> Self {
+        let mem_usage = meter.inner.gauge_f64("resource_slots_mem_usage".into());
+        let cpu_usage = meter.inner.gauge_f64("resource_slots_cpu_usage".into());
+        let mem_pid_output = meter
+            .inner
+            .gauge_f64("resource_slots_mem_pid_output".into());
+        let cpu_pid_output = meter
+            .inner
+            .gauge_f64("resource_slots_cpu_pid_output".into());
+        let attribs = meter.inner.new_attributes(meter.default_attribs);
+        Self {
+            attribs,
+            mem_usage,
+            cpu_usage,
+            mem_pid_output,
+            cpu_pid_output,
+        }
+    }
+}
+
+/// Implementors provide information about system resource usage
+pub trait SystemResourceInfo {
+    /// Return total available system memory in bytes
+    fn total_mem(&self) -> u64;
+    /// Return memory used by the system in bytes
+    fn used_mem(&self) -> u64;
+    /// Return system used CPU as a float in the range [0.0, 1.0] where 1.0 is defined as all
+    /// cores pegged
+    fn used_cpu_percent(&self) -> f64;
+    /// Return system used memory as a float in the range [0.0, 1.0]
+    fn process_used_percent(&self) -> f64 {
+        self.used_mem() as f64 / self.total_mem() as f64
+    }
+}
+
+#[async_trait::async_trait]
+impl<MI, SK> SlotSupplier for ResourceBasedSlotsForType<MI, SK>
+where
+    MI: SystemResourceInfo + Send + Sync,
+    SK: SlotKind + Send + Sync,
+{
+    type SlotKind = SK;
+
+    async fn reserve_slot(&self, ctx: &dyn SlotReservationContext) -> SlotSupplierPermit {
+        loop {
+            // TODO: Just using this for now to make sure metrics get emitted regularly
+            self.pid_decision();
```

Comment:
I still need to deal with this, but it won't affect much. Likely just a spawned task to periodically emit them.